RationaleMCP/0027/HindleyMilner.md

```diff
@@ -0,0 +1,201 @@
+# Unit checking
+Unit checking is performed by using a type inference algorithm, albeit adapted to work with units. An example of such an algorithm is the Hindley-Milner algorithm.
+At a high level the algorithm collects a set of constraints on unit of expressions. It extracts a sequence of substitutions from these constraints which once applied constructs the unit of a given expression.
+Any inconsistencies discovered during this process, is a unit error, and any unit expression that can't be evaluated is considered undefined.
+
+## Unit of expressions
+
+The unit of an arbitrary Modelica expression is represented by a unit meta-expression.
+This sections outlines the different constructs it comprises.
+
+### Variables
+The unit of variable `var`, is represented by identifier `var.unit`.
+
+### Literals
+
+#### `Well-formed` unit
+A `well-formed` unit can be uniquely represented by its base units factorization, offset and scaling, up to the order of the base units.
+
+#### `Empty` unit
+An expression with `empty` unit won't contribute to the unit of a parent expression, and will meet any constraints imposed on it.
+
+#### `Undefined` unit
+An expression with `undefined` unit is used to represent unspecified cases, and will meet any constraints imposed on it.
+
+### Operators
+
+#### div
+Consider the meta-expression `e1 div e2`.
+* If both `e1` and `e2` are `well-formed`, it evaluates to a `well-formed` unit resulting from dividing `e1` by `e2`.
+* If both `e1` and `e2` are `empty`, it evaluates to `empty`.
+* If `e1` is `empty` and `e2` is well-formed, it evaluates to a `well-formed` unit resulting from dividing `"1"` by `e2`.
+* If `e1` is `well-formed` and `e2` is `empty`, it evaluates to `e1`.
+* If one of them is `undefined`, it evaluates to `undefined`.
+
+#### mul
+Consider the meta-expression `e1 mul e2`.
+* If both `e1` and `e2` are `well-formed`, it evaluates to a `well-formed` unit resulting from multiplying `e1` by `e2`.
+* If both `e1` and `e2` are `empty`, it evaluates to `empty`.
+* If one of them is `empty`, it evaluates to the other one.
+* If one of them is `undefined`, it evaluates to `undefined`.
+
+#### pow
+Consider the meta-expression `b pow e`.
+* If `b` is `well-formed` and `e` is an Integer, it evaluates to a `well-formed` unit resulting from raising `b` to the power `e`.
```

Comment:
I think there's a mix of Modelica expression and meta-expression going on here.  In a meta-expression, I would have expected the exponent `e` to always be a rational number. Am I missing something?

Reply:
Maybe one needs a separate section on the unit semantics of (a selection of) Modelica operators too?

Reply:
> I think there's a mix of Modelica expression and meta-expression going on here. In a meta-expression, I would have expected the exponent `e` to always be a rational number. Am I missing something?

If the wellformed unit is `"1"` we should still accept `Real` exponents.

Reply:
> Maybe one needs a separate section on the unit semantics of (a selection of) Modelica operators too?

At the very least, this is what I tried to do in the `Unit Inference` section.

Reply:
Sorry, I forgot that there was also the later section!  I forgot about it just because I saw unexpected things here.  Wouldn't it make sense to move the case analysis to the later section, so that what is described here is only a meta-expression raised to a rational?